src/query/sql/src/executor/physical_plan_builder.rs

```diff
@@ -855,8 +862,99 @@ impl PhysicalPlanBuilder {
+                            ty.clone(),
+                            &BUILTIN_FUNCTIONS.default_cast_rules,
+                        )
+                        .ok_or_else(|| {
+                            ErrorCode::IllegalDataType(format!(
+                                "Cannot find common type for {:?} and {:?}",
+                                &common_ty, &ty
+                            ))
+                        })?;
+                    }
+
+                    *order_by = wrap_cast(order_by, &common_ty);
+                    for scalar in start.iter_mut().chain(end.iter_mut()) {
+                        let raw_expr = RawExpr::<usize>::Cast {
+                            span: w.span,
+                            is_try: false,
+                            expr: Box::new(RawExpr::Constant {
+                                span: w.span,
+                                scalar: scalar.clone(),
+                            }),
+                            dest_type: common_ty.clone(),
+                        };
+                        let expr = type_check::check(&raw_expr, &BUILTIN_FUNCTIONS)?;
+                        let (expr, _) = ConstantFolder::fold(
+                            &expr,
+                            &FunctionContext::default(),
+                            &BUILTIN_FUNCTIONS,
+                        );
+                        if let common_expression::Expr::Constant {
+                            scalar: new_scalar, ..
+                        } = expr
+                        {
+                            if new_scalar.is_positive() {
+                                **scalar = new_scalar;
+                                continue;
+                            }
+                        }
+                        return Err(ErrorCode::SemanticError(
+                            "Only positive numbers are allowed in RANGE offset".to_string(),
+                        )
+                        .set_span(w.span));
+                    }
+                }
+
+                // Generate a `EvalScalar` as the input of `Window`.
+                let mut scalar_items: Vec<ScalarItem> = Vec::new();
+                for arg in &w.arguments {
+                    scalar_items.push(arg.clone());
+                }
+                for part in &w.partition_by {
+                    scalar_items.push(part.clone());
+                }
+                for order in &w.order_by {
+                    scalar_items.push(order.order_by_item.clone())
+                }
+                let input = if !scalar_items.is_empty() {
+                    let eval_scalar = crate::planner::plans::EvalScalar {
+                        items: scalar_items,
+                    };
+                    SExpr::create_unary(eval_scalar.into(), s_expr.child(0)?.clone())
+                } else {
+                    s_expr.child(0)?.clone()
+                };
+                let input = self.build(&input).await?;
```

Comment:
I think we can add `EvalScalar` plan manually instead of `build` again. (Avoid traversing the plan tree twice)

![image](https://github.com/datafuselabs/databend/assets/33975039/e7150113-5d5e-4836-a597-94f333e0b066)

This input plan can be the child of the new `EvalScalar`, and we can reuse the codes for building `EvalScalar`.

Reply:
The duplicate build is to avoid repeating the code of building physical EvalScalar

Reply:
But it will traverse the plan again. I think we can avoid this.